### PR TITLE
feat: [CDS-74007]: matrix not working as expected

### PR DESCRIPTION
**Pull request by sahilhindwani-harness**
## Describe your changes
When the matrix axis was defined as an expression of a harness object, then recaster was added into it. Because of that recaster, we were converting that matrix map to a java pojo and we had that handling already to handle map but not these pojos.

As part of this fix, added a check for java objects too.

Validated on local and also added a test

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50338)

### CODE REVIEW

File: 120-ng-manager/config.yml [hunk 1]
```diff
@@ -277,7 +277,7 @@ pmsGrpcClientConfig:
   target: localhost:12011
   authority: localhost
 
-shouldConfigureWithPMS: ${SHOULD_CONFIGURE_WITH_PMS:-false}
+shouldConfigureWithPMS: ${SHOULD_CONFIGURE_WITH_PMS:-true}
```
Comment (naidu-harness) [inline 120-ng-manager/config.yml:280]: revert this file change

**Review by naidu-harness — CHANGES_REQUESTED** [review]

**Review by archit-harness — APPROVED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50338)* status: 4 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Comment by sahilhindwani-harness** [comment]
trigger uts

**Comment by sahilhindwani-harness** [comment]
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java` line 311 at r1](https://reviewable.io/reviews/harness/harness-core/50338#gh-1257900720:-N_8VOnD1hUr4qcay-Mf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/01ef71c7bbca58b51e57c112420e36eeb22e2f7b/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java#L311)):*
<details><summary><i>Previously, naidu-harness (Sanyasi Naidu Annepu) wrote…</i></summary><blockquote>

Lets provide the Key/Val in response
</blockquote></details>

Done.

**Comment by sahilhindwani-harness** [comment]
*[`120-ng-manager/config.yml` line 280 at r1](https://reviewable.io/reviews/harness/harness-core/50338#gh-1257901182:-N_8VPiN9TmZ57j4E4gU:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/01ef71c7bbca58b51e57c112420e36eeb22e2f7b/120-ng-manager/config.yml#L280)):*
<details><summary><i>Previously, naidu-harness (Sanyasi Naidu Annepu) wrote…</i></summary><blockquote>

revert this file change
</blockquote></details>

Done.

**Pull request merged**